streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java

```diff
@@ -227,21 +253,72 @@ public final TopologyBuilder addSource(String name, Deserializer keyDeserializer
         if (nodeFactories.containsKey(name))
             throw new TopologyBuilderException("Processor " + name + " is already added.");
 
+        if (topicPattern != null) {
+            throw new TopologyBuilderException("Subscription to topics, partitions and pattern are mutually exclusive");
+        }
+
         for (String topic : topics) {
             if (sourceTopicNames.contains(topic))
                 throw new TopologyBuilderException("Topic " + topic + " has already been registered by another source.");
 
             sourceTopicNames.add(topic);
         }
 
-        nodeFactories.put(name, new SourceNodeFactory(name, topics, keyDeserializer, valDeserializer));
+        nodeFactories.put(name, new SourceNodeFactory(name, topics, null, keyDeserializer, valDeserializer));
         nodeToSourceTopics.put(name, topics.clone());
         nodeGrouper.add(name);
 
         return this;
     }
 
     /**
+     * Add a new source that consumes from topics matching the given pattern
```

Comment:
We should probably mention that due to consumer design restriction, currently we only allow one stream throughout the topology to be created from regex patterns.